File: 2022/day01/main.ts
#!/usr/bin/env deno run --allow-read
import { assertEquals } from "https://deno.land/std@0.116.0/testing/asserts.ts";
import { sum } from "../../2020/utils.ts";

const parseInput = (string: string): number[][] =>
  string.trim().split("\n\n").map((block) => {
    const lines = block.trim().split("\n");
    return lines.map((line) => parseInt(line.trim(), 10));
  });

const text = await Deno.readTextFile("input.txt");

const input = parseInput(text);

const part1 = (input: number[][]): number => {
  const sums = input.map(sum);
  return Math.max(...sums);
};

const example = parseInput(`
  1000
  2000
  3000

  4000

  5000
  6000

  7000
  8000
  9000

  10000
`);

assertEquals(part1(example), 24000);

console.log("Result part 1: " + part1(input));

const part2 = (input: number[][]): number => {
  const sums = input.map(sum);
  const sortedSums = sums.sort((a, b) => b - a);
  const topThree = sortedSums.slice(0, 3);
  return sum(topThree);
};

assertEquals(part2(example), 45000);

console.log("Result part 2: " + part2(input));

File: 2022/day01/main.rs

fn parse_block(block: &str) -> Vec<u32> {
    let pieces = block.trim().split("\n");
    pieces.map(|line| line.parse::<u32>().unwrap()).collect()
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = std::fs::read_to_string("day01/input.txt")?;

    let parsed_input  = file
        .split("\n\n")
        .map(|block| parse_block(block)).collect::<Vec<Vec<u32>>>();

    let mut sums = parsed_input.iter().map(|block| block.iter().sum()).collect::<Vec<u32>>();

    let part1 = sums.iter().max().unwrap();

    println!("part 1: {}", part1);

    sums.sort();

    let part2 = sums[sums.len()-3..sums.len()].iter().sum::<u32>();

    println!("part 2: {}", part2);

    Ok(())
}
